wgpu-hal/src/gles/web.rs

```diff
@@ -28,6 +28,38 @@ pub struct Instance {
     canvas: Mutex<Option<web_sys::HtmlCanvasElement>>,
 }
 
+impl Instance {
+    pub fn create_surface_from_canvas(
+        &self,
+        canvas: &web_sys::HtmlCanvasElement,
+    ) -> Result<Surface, crate::InstanceError> {
+        *self.canvas.lock() = Some(canvas.clone());
+
+        Ok(Surface {
+            canvas: canvas.clone(),
+            present_program: None,
+            swapchain: None,
+            texture: None,
+            presentable: true,
+        })
+    }
+
+    pub fn create_surface_from_offscreen_canvas(
+        &self,
+        canvas: &web_sys::OffscreenCanvas,
+    ) -> Result<Surface, crate::InstanceError> {
+        *self.canvas.lock() = Some(canvas.clone());
```

Comment:
@zicklag Do you think we could create the `webgl2_context` already when creating the surface and keep it instead of the canvas? This way, it would be more similar to the [WebGPU implementation](https://github.com/gfx-rs/wgpu/blob/cb7ad30d585ea68ae9a2652ad5251ef711cbbf94/wgpu/src/backend/web.rs#L953) I think.
Otherwise, I would probably need to create some kind of CanvasWrapper that holds the Canvas or OffscreenCanvas...

Reply:
I think we could do that. Then we would have to put the context inside the mutex on the `Instance` instead of the canvas, so that we could get to the context when we `enumerate_adapters()`.

Then, like you say, we wouldn't have to have an enum like `enum CanvasType { Normal(Canvas), Offscreen(OffscreenCanvas) }`.

Haven't thought of any problems with that yet. Try it. :) :+1: